qiskit_addon_utils/exp_vals/expectation_values.py

```diff
@@ -0,0 +1,255 @@
+            At least one more axis must be defined indicating the measurement bases, at position `meas_basis_axis` and of length `len(basis_dict)`.
+        basis_dict: This dict encodes how the data in `bool_array` should be used to estimate the desired list of Pauli observables.
+            The ith key is a measurement basis assumed to correspond to the ith slice of `bool_array` along the `meas_basis_axis` axis.
+            Each dict value is a list of length equal to the number of desired observables.
+            The jth element of this list is a `SparsePauliOp` assumed to be compatible (qubit-wise commuting) with the measurement-basis key.
+            In place of a `SparsePauliOp`, `None` may be used to represent the zero operator, when a basis is not used to compute an observable.
+            The jth observable is defined as the sum of the jth element of each dict value (contribution from each meas basis).
+            - Note the order of dict entries is relied on here for indexing; the dict keys are never used.
+            - Assumes each Pauli term (in dict values) is compatible with each measurement basis (in keys).
+            - Assumes each term in each observable appears for exactly one basis (TODO: remove this assumption).
+        meas_basis_axis: Axis of bool_array that indexes measurement bases. Ordering must match ordering in `basis_dict`.
+        avg_axis: Optional axis or axes of bool_array to average over when computing expectation values. Usually this is the "twirling" axis.
+            Must be nonnegative. (The shots axis, assumed to be at index -2 in the boolean array, is always averaged over).
+        meas_flips: Optional boolean array used with measurement twirling. Indicates which bits were acquired with measurements preceded by bit-flip gates.
+            Data processing will use the result of `xor`ing this array with `bool_array`. Must be same shape as `bool_array`.
+        pec_signs: Optional boolean array used with probabilistic error cancellation (PEC). Indicates which errors were inserted in each
+            circuit randomization. Final axis, assumed to index all error generators in circuit, is immediately collapsed as a sum mod 2.
+            Remaining shape must be `pec_signs.shape[:-1] == bool_array.shape[:-2]`. Note this array does not have a shots axis.
+        postselect_mask: Optional boolean array used for postselection. `True` (`False`) indicates a shot accepted (rejected) by postselection.
+            Shape must be `bool_array.shape[:-1]`.
+
+    Returns:
+        A list of (exp. val, variance) 2-tuples, one for each desired observable.
+    """
+    if avg_axis is None:
+        avg_axis = tuple()
+    elif isinstance(avg_axis, int):
+        avg_axis = (avg_axis,)
+    else:
+        avg_axis = tuple(avg_axis)
+
+    if any(a < 0 for a in avg_axis):
+        raise ValueError("`avg_axis` must be nonnegative")
```

Comment:
Just curious: why is this a requirement?

Reply:
It's somewhat ambiguous whether negative indexing should include the shots and bits axes when counting backwards from the last axis, so mainly avoiding that potential confusion.